server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregatorFactory.java

```diff
@@ -61,24 +64,35 @@ public long minDocCount() {
     }
 
     @Override
-    protected Aggregator doCreateInternal(ValuesSource.Numeric valuesSource, Aggregator parent, boolean collectsFromSingleBucket,
+    protected Aggregator doCreateInternal(ValuesSource valuesSource, Aggregator parent, boolean collectsFromSingleBucket,
             List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
         if (collectsFromSingleBucket == false) {
             return asMultiBucketAggregator(this, context, parent);
         }
-        return createAggregator(valuesSource, parent, pipelineAggregators, metaData);
-    }
-
-    private Aggregator createAggregator(ValuesSource.Numeric valuesSource, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
-            Map<String, Object> metaData) throws IOException {
-
-        return new HistogramAggregator(name, factories, interval, offset, order, keyed, minDocCount, minBound, maxBound, valuesSource,
-                config.format(), context, parent, pipelineAggregators, metaData);
+        if (valuesSource instanceof ValuesSource.Numeric) {
+            return new NumericHistogramAggregator(name, factories, interval, offset, order, keyed, minDocCount, minBound, maxBound,
+                (ValuesSource.Numeric) valuesSource, config.format(), context, parent, pipelineAggregators, metaData);
+        }
+        else if (valuesSource instanceof ValuesSource.Bytes.FieldData.RangeFieldData) {
+            ValuesSource.Bytes.FieldData.RangeFieldData rangeValueSource = (ValuesSource.Bytes.FieldData.RangeFieldData) valuesSource;
+            if (rangeValueSource.rangeType().isNumeric() == false) {
+                throw new IllegalArgumentException("Expected numeric range type but found non-numeric range ["
+                    + rangeValueSource.rangeType().name + "]");
+            }
+            return new RangeHistogramAggregator(name, factories, interval, offset, order, keyed, minDocCount, minBound, maxBound,
+                (ValuesSource.Bytes.FieldData.RangeFieldData) valuesSource, config.format(), context, parent, pipelineAggregators,
+                metaData);
+        }
+        else {
+            throw new IllegalArgumentException("Expected one of [Numeric, RangeFieldData] values source, found ["
+                + valuesSource.toString() + "]");
+        }
     }
 
     @Override
     protected Aggregator createUnmapped(Aggregator parent, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
             throws IOException {
-        return createAggregator(null, parent, pipelineAggregators, metaData);
+        return new NumericHistogramAggregator(name, factories, interval, offset, order, keyed, minDocCount, minBound, maxBound,
+            null, config.format(), context, parent, pipelineAggregators, metaData);
```

Comment:
I'm almost positive this is the wrong thing to do here, but I'm not sure what the right thing to do is.

Reply:
Yeah, so this is a bit of a split personality in the framework right now.  Setting the valuesSource to null will use a no-op collector in the histo's `getLeafCollector()`.  Then it uses the same histo aggregator to build an empty result.

Terms agg (and a few others) instead have an "unmapped" version of the agg (`UnmappedTerms`) which is used instead.  This unmapped version uses the same no-op collector, and knows how to build an empty response.

I'm not sure if there is a preference or why it's a split-personality.  I've pinged @colings86 to see if he knows the history here :)